app/agent/git/git.service.ts

```diff
@@ -286,40 +300,56 @@ export class AgentGitService {
         privateSshKey: string;
       };
-    },
-  ): Promise<void> {
-    const outputDriveExists = await doesFileExist(outputDrive.pathOnHost);
-    if (!outputDriveExists) {
-      await fs.mkdir(path.dirname(outputDrive.pathOnHost), { recursive: true });
-      await this.agentUtilService.createExt4Image(outputDrive.pathOnHost, outputDrive.maxSizeMiB);
-      this.logger.info(`empty output image created at ${outputDrive.pathOnHost}`);
+    };
+  }): Promise<void> {
+    const { runtime, outputId, repository, tmpContentPrefix } = args;
+    const localRootFsImageTag = `fetchrepo-${sha256(this.agentConfig.fetchRepoImageTag)}`;
+    const rootFsImageId = BlockRegistryService.genImageId(localRootFsImageTag);
+    if (!(await this.blockRegistryService.hasContent(rootFsImageId))) {
+      await this.blockRegistryService.loadImageFromRemoteRepo(
+        this.agentConfig.fetchRepoImageTag,
+        localRootFsImageTag,
+      );
     }
-    const outputDir = "/tmp/output";
-    return await withFileLock(outputDrive.pathOnHost, async () => {
-      await firecrackerService.withVM(
+    const rootFsContainerId = await this.blockRegistryService.createContainer(
+      rootFsImageId,
+      `${tmpContentPrefix}-fetchrepo-${uuidv4()}`,
+    );
+    const containerId = BlockRegistryService.genContainerId(outputId);
+    return await withLocalLock(LocalLockNamespace.CONTAINER, containerId, async () => {
+      await this.blockRegistryService.createContainer(void 0, outputId, {
+        mkfs: true,
+        sizeInGB: 64,
+      });
+      const outputDir = "/fr";
+      await withRuntimeAndImages(
+        this.blockRegistryService,
+        runtime,
         {
           ssh: {
             username: "hocus",
             password: "hocus",
           },
           kernelPath: this.agentConfig.defaultKernel,
-          rootFsPath: this.agentConfig.fetchRepositoryRootFs,
-          copyRootFs: true,
-          extraDrives: [
-            {
-              pathOnHost: outputDrive.pathOnHost,
-              guestMountPath: outputDir,
-            },
-          ],
-          memSizeMib: 4096,
-          vcpuCount: 2,
+          fs: {
+            "/": rootFsContainerId,
+            [outputDir]: containerId,
+          },
+          memSizeMib: 1024,
+          vcpuCount: 1,
         },
         async ({ ssh }) => {
           const repositoryDir = path.join(outputDir, PROJECT_DIR);
-          if (!outputDriveExists) {
+          const outputDirOwner = await execSshCmd({ ssh }, [
+            "sudo",
+            "stat",
+            "-c",
+            "%U",
+            outputDir,
+          ]).then((r) => r.stdout.trim());
+          if (outputDirOwner !== "hocus") {
             await execSshCmd({ ssh }, ["sudo", "chown", "-R", "hocus:hocus", outputDir]);
           }
-
           const sshKey = repository.credentials.privateSshKey;
           const sshDir = "/home/hocus/.ssh";
           const sshKeyPath = path.join(sshDir, `${uuidv4()}.key`);
```

Comment:
Please write an TODO here to use ssh agent forwarding so we're not C&P priv keys between vms

Reply:
Why not just copy and paste?